File: src/main/java/aurocosh/divinefavor/common/item/talismans/spell/SpellTalismanEarthenDive.java
package aurocosh.divinefavor.common.item.talismans.spell;

import aurocosh.divinefavor.common.config.common.ConfigSpells;
import aurocosh.divinefavor.common.favor.ModFavor;
import aurocosh.divinefavor.common.item.talismans.spell.base.ItemSpellTalisman;
import aurocosh.divinefavor.common.item.talismans.spell.base.SpellOptions;
import aurocosh.divinefavor.common.item.talismans.spell.base.TalismanContext;
import aurocosh.divinefavor.common.util.UtilCoordinates;
import aurocosh.divinefavor.common.util.UtilEntity;
import net.minecraft.util.math.BlockPos;

import java.util.EnumSet;

public class SpellTalismanEarthenDive extends ItemSpellTalisman {
    public SpellTalismanEarthenDive(String name, ModFavor favor, int favorCost, EnumSet<SpellOptions> options) {
        super(name, favor, favorCost, options);
    }

    @Override
    protected void performActionServer(TalismanContext context) {
        BlockPos targetPos = UtilCoordinates.findPlaceToStandBelow(context.pos, context.world, ConfigSpells.earthenDive.maxDistance, true);
        if (targetPos != null)
            UtilEntity.teleport(context.player, targetPos);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/spell_talismans/SpellTalismanEarthenDive.kt
package aurocosh.divinefavor.common.item.spell_talismans

import aurocosh.divinefavor.common.config.common.ConfigSpell
import aurocosh.divinefavor.common.item.spell_talismans.base.ItemSpellTalisman
import aurocosh.divinefavor.common.item.spell_talismans.base.SpellOptions
import aurocosh.divinefavor.common.item.spell_talismans.context.CastContext
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.util.UtilCoordinates
import aurocosh.divinefavor.common.util.UtilEntity
import net.minecraft.block.material.Material
import java.util.*

class SpellTalismanEarthenDive(name: String, spirit: ModSpirit, favorCost: Int, options: EnumSet<SpellOptions>) : ItemSpellTalisman(name, spirit, favorCost, options) {

    override fun performActionServer(context: CastContext) {
        val world = context.world
        val targetPos = UtilCoordinates.findPlaceToStandBelow(context.pos, world, ConfigSpell.earthenDive.maxDistance, true) ?: return

        val lavaIsClose = UtilCoordinates.getBlocksInSphere(targetPos, 2)
                .map(world::getBlockState)
                .any { it.material == Material.LAVA }
        if (!lavaIsClose)
            UtilEntity.teleport(context.player, targetPos)
    }
}
